File: C++ Playground/05_TwoPointers.cpp
/// Two Pointers Playground
/// Opposite direction for sorted/palindrome problems; same direction for
/// in-place removal/partition; sort first for k-sum; use while guards to
/// skip duplicates; avoid off-by-one in bounds.

#include <algorithm>
#include <cctype>
#include <iostream>
#include <vector>
#include <string>
#include <string_view>

namespace two_pointers_playground {

// ============================================================================
// OPPOSITE DIRECTION
// ============================================================================

/// @brief Find two numbers that add up to target in sorted array.
/// @param nums Sorted array of integers
/// @param target Sum to find
/// @return [index1, index2] 1-indexed, index1 < index2
/// @example TwoSumSorted({2,7,11,15}, 9) -> {1,2}
/// @constraints Array is sorted; exactly one solution
/// @time O(n) @space O(1)
[[nodiscard]] std::vector<int>
TwoSumSorted(const std::vector<int>& nums, int target) noexcept {
  int left = 0;
  int right = static_cast<int>(nums.size()) - 1;
  while (left < right) {
    int sum = nums[left] + nums[right];
    if (sum == target) {
      return {left + 1, right + 1};
    }
    if (sum < target) {
      ++left;
    } else {
      --right;
    }
  }
  return {};
}

/// @brief Find all unique triplets summing to target.
/// @param nums Array of integers
/// @param target Sum to find
/// @return Vector of unique triplets
/// @example ThreeSum({-1,0,1,2,-1,-4}, 0) -> {[-1,-1,2], [-1,0,1]}
/// @constraints O(n²) time required; handle duplicates
/// @time O(n²) @space O(1) excluding output
/// Tricky: Sort first; skip duplicates carefully in outer loop
[[nodiscard]] std::vector<std::vector<int>>
ThreeSum(std::vector<int> nums) noexcept {
  std::vector<std::vector<int>> result;
  std::sort(nums.begin(), nums.end());

  for (size_t i = 0; i < nums.size(); ++i) {
    if (nums[i] > 0) break;
    if (i > 0 && nums[i] == nums[i - 1]) continue;

    int left = i + 1;
    int right = nums.size() - 1;
    int target = -nums[i];

    while (left < right) {
      int sum = nums[left] + nums[right];
      if (sum == target) {
        result.emplace_back(
            std::vector<int>{nums[i], nums[left], nums[right]});
        while (left < right && nums[left] == nums[left + 1]) ++left;
        while (left < right && nums[right] == nums[right - 1]) --right;
        ++left;
        --right;
      } else if (sum < target) {
        ++left;
      } else {
        --right;
      }
    }
  }
  return result;
}

/// @brief Find all unique quadruplets summing to target.
/// @param nums Array of integers
/// @param target Sum to find
/// @return Vector of unique quadruplets
/// @example FourSum({1000000000,1000000000,1000000000,1000000000}, -294967296)
/// @constraints Handle overflow; two nested sorted iterations
/// @time O(n³) @space O(1) excluding output
/// Tricky: Use long long to avoid integer overflow in sum
[[nodiscard]] std::vector<std::vector<int>>
FourSum(std::vector<int> nums, long long target) noexcept {
  std::vector<std::vector<int>> result;
  if (nums.size() < 4) return result;

  std::sort(nums.begin(), nums.end());

  for (size_t i = 0; i < nums.size() - 3; ++i) {
    if (i > 0 && nums[i] == nums[i - 1]) continue;
    long long first = nums[i];
    if (first + nums[i + 1] + nums[i + 2] + nums[i + 3] > target) break;
    if (first + nums[nums.size() - 3] + nums[nums.size() - 2] +
            nums[nums.size() - 1] <
        target)
      continue;

    for (size_t j = i + 1; j < nums.size() - 2; ++j) {
      if (j > i + 1 && nums[j] == nums[j - 1]) continue;
      long long second = nums[j];
      if (first + second + nums[j + 1] + nums[j + 2] > target) break;
      if (first + second + nums[nums.size() - 2] + nums[nums.size() - 1] <
          target)
        continue;

      int left = j + 1;
      int right = nums.size() - 1;
      while (left < right) {
        long long sum = first + second + nums[left] + nums[right];
        if (sum == target) {
          result.emplace_back(std::vector<int>{
              nums[i], nums[j], nums[left], nums[right]});
          while (left < right && nums[left] == nums[left + 1]) ++left;
          while (left < right && nums[right] == nums[right - 1]) --right;
          ++left;
          --right;
        } else if (sum < target) {
          ++left;
        } else {
          --right;
        }
      }
    }
  }
  return result;
}

/// @brief Maximum water area between two lines.
/// @param height Heights at each index
/// @return Maximum container area
/// @example ContainerWithMostWater({1,8,6,2,5,4,8,3,7}) -> 49
/// @constraints Height >= 0
/// @time O(n) @space O(1)
/// Tricky: Area = min(h[i], h[j]) * (j-i); move inward from smaller
[[nodiscard]] int
ContainerWithMostWater(const std::vector<int>& height) noexcept {
  int left = 0;
  int right = height.size() - 1;
  int max_area = 0;

  while (left < right) {
    int width = right - left;
    int h = std::min(height[left], height[right]);
    int area = width * h;
    max_area = std::max(max_area, area);

    if (height[left] < height[right]) {
      ++left;
    } else {
      --right;
    }
  }
  return max_area;
}

// ============================================================================
// SAME DIRECTION
// ============================================================================

/// @brief Remove duplicates in-place from sorted array.
/// @param nums Sorted array, modified in-place
/// @return Number of unique elements
/// @example RemoveDuplicates({1,1,2}) -> 2, nums becomes {1,2,...}
/// @constraints Array is sorted; at most one instance retained per value
/// @time O(n) @space O(1)
[[nodiscard]] int RemoveDuplicates(std::vector<int>& nums) noexcept {
  if (nums.empty()) return 0;

  int write = 0;
  for (int read = 1; read < static_cast<int>(nums.size()); ++read) {
    if (nums[read] != nums[write]) {
      ++write;
      nums[write] = nums[read];
    }
  }
  return write + 1;
}

/// @brief Remove all instances of val in-place.
/// @param nums Array to modify
/// @param val Value to remove
/// @return Count of remaining elements
/// @example RemoveElement({0,1,2,2,3,0,4}, 2) -> 5
/// @constraints Order of remaining elements doesn't matter
/// @time O(n) @space O(1)
[[nodiscard]] int RemoveElement(std::vector<int>& nums,
                                 int val) noexcept {
  int write = 0;
  for (int read = 0; read < static_cast<int>(nums.size()); ++read) {
    if (nums[read] != val) {
      nums[write] = nums[read];
      ++write;
    }
  }
  return write;
}

/// @brief Move all zeros to end, preserve relative order of non-zero.
/// @param nums Array to modify in-place
/// @example MoveZeroes({0,1,0,3,12}) -> {1,3,12,0,0}
/// @constraints O(n) time, O(1) space required
/// @time O(n) @space O(1)
void MoveZeroes(std::vector<int>& nums) noexcept {
  int write = 0;
  for (int read = 0; read < static_cast<int>(nums.size()); ++read) {
    if (nums[read] != 0) {
      if (write != read) {
        nums[write] = nums[read];
      }
      ++write;
    }
  }
  while (write < static_cast<int>(nums.size())) {
    nums[write] = 0;
    ++write;
  }
}

// ============================================================================
// PARTITION (THREE-WAY)
// ============================================================================

/// @brief Sort array with 0, 1, 2 in-place (Dutch National Flag).
/// @param nums Array containing only 0, 1, 2; modified in-place
/// @example SortColors({2,0,1}) -> {0,1,2}
/// @constraints One pass required; O(1) space
/// @time O(n) @space O(1)
/// Tricky: Maintain three pointers: low (0 boundary), mid (current),
/// high (2 boundary)
void SortColors(std::vector<int>& nums) noexcept {
  int low = 0;
  int mid = 0;
  int high = nums.size() - 1;

  while (mid <= high) {
    if (nums[mid] == 0) {
      std::swap(nums[low], nums[mid]);
      ++low;
      ++mid;
    } else if (nums[mid] == 1) {
      ++mid;
    } else {
      std::swap(nums[mid], nums[high]);
      --high;
    }
  }
}

// ============================================================================
// PALINDROME
// ============================================================================

/// @brief Check if string is palindrome ignoring non-alphanumeric.
/// @param s Input string
/// @return true if palindrome (case-insensitive alphanumeric only)
/// @example ValidPalindrome("A man, a plan, a canal: Panama") -> true
/// @constraints O(n) time, O(1) space required
/// @time O(n) @space O(1)
[[nodiscard]] bool ValidPalindrome(std::string_view s) noexcept {
  int left = 0;
  int right = s.size() - 1;

  while (left < right) {
    while (left < right && !std::isalnum(s[left])) {
      ++left;
    }
    while (left < right && !std::isalnum(s[right])) {
      --right;
    }
    if (std::tolower(s[left]) != std::tolower(s[right])) {
      return false;
    }
    ++left;
    --right;
  }
  return true;
}

/// @brief Check if palindrome with at most one character deletion allowed.
/// @param s Input string
/// @return true if can be palindrome by deleting at most one char
/// @example ValidPalindromeII("abca") -> true (delete 'b')
/// @constraints Helper checks palindrome range
/// @time O(n) @space O(1)
[[nodiscard]] bool ValidPalindromeII(std::string_view s) noexcept {
  auto is_palindrome = [](std::string_view str, int left,
                          int right) noexcept {
    while (left < right) {
      if (str[left] != str[right]) return false;
      ++left;
      --right;
    }
    return true;
  };

  int left = 0;
  int right = s.size() - 1;

  while (left < right) {
    if (s[left] != s[right]) {
      return is_palindrome(s, left + 1, right) ||
             is_palindrome(s, left, right - 1);
    }
    ++left;
    --right;
  }
  return true;
}

// ============================================================================
// RAIN WATER TRAPPING
// ============================================================================

/// @brief Calculate trapped rainwater given elevation map.
/// @param height Elevation at each position
/// @return Total water trapped
/// @example Trap({0,1,0,2,1,0,1,3,2,1,2,1}) -> 6
/// @constraints O(1) space solution; track left/right max
/// @time O(n) @space O(1)
/// Tricky: Process from side with smaller max; water level = min(left_max,
/// right_max) - height[i]
[[nodiscard]] int Trap(const std::vector<int>& height) noexcept {
  if (height.empty()) return 0;

  int left = 0;
  int right = height.size() - 1;
  int left_max = 0;
  int right_max = 0;
  int water = 0;

  while (left < right) {
    if (height[left] < height[right]) {
      if (height[left] >= left_max) {
        left_max = height[left];
      } else {
        water += left_max - height[left];
      }
      ++left;
    } else {
      if (height[right] >= right_max) {
        right_max = height[right];
      } else {
        water += right_max - height[right];
      }
      --right;
    }
  }
  return water;
}

}  // namespace two_pointers_playground

// ============================================================================
// TESTS
// ============================================================================

int main() {
  using namespace two_pointers_playground;

  std::cout << "=== Two Pointers Playground ===\n\n";

  std::cout << "TwoSumSorted({2,7,11,15}, 9): ";
  {
    auto result = TwoSumSorted({2, 7, 11, 15}, 9);
    std::cout << "[" << result[0] << ", " << result[1] << "]\n";
  }

  std::cout << "ThreeSum({-1,0,1,2,-1,-4}, 0): ";
  {
    auto result = ThreeSum({-1, 0, 1, 2, -1, -4});
    for (const auto& triplet : result) {
      std::cout << "[" << triplet[0] << "," << triplet[1] << ","
                << triplet[2] << "] ";
    }
    std::cout << "\n";
  }

  std::cout << "ContainerWithMostWater({1,8,6,2,5,4,8,3,7}): ";
  {
    auto result = ContainerWithMostWater({1, 8, 6, 2, 5, 4, 8, 3, 7});
    std::cout << result << "\n";
  }

  std::cout << "RemoveDuplicates({1,1,2}): ";
  {
    std::vector<int> nums = {1, 1, 2};
    int len = RemoveDuplicates(nums);
    std::cout << "len=" << len << "\n";
  }

  std::cout << "RemoveElement({0,1,2,2,3,0,4}, 2): ";
  {
    std::vector<int> nums = {0, 1, 2, 2, 3, 0, 4};
    int len = RemoveElement(nums, 2);
    std::cout << "len=" << len << "\n";
  }

  std::cout << "MoveZeroes({0,1,0,3,12}): ";
  {
    std::vector<int> nums = {0, 1, 0, 3, 12};
    MoveZeroes(nums);
    for (int x : nums) std::cout << x << " ";
    std::cout << "\n";
  }

  std::cout << "SortColors({2,0,1}): ";
  {
    std::vector<int> nums = {2, 0, 1};
    SortColors(nums);
    for (int x : nums) std::cout << x << " ";
    std::cout << "\n";
  }

  std::cout << "ValidPalindrome(\"A man, a plan, a canal: Panama\"): ";
  {
    bool result = ValidPalindrome("A man, a plan, a canal: Panama");
    std::cout << (result ? "true" : "false") << "\n";
  }

  std::cout << "ValidPalindromeII(\"abca\"): ";
  {
    bool result = ValidPalindromeII("abca");
    std::cout << (result ? "true" : "false") << "\n";
  }

  std::cout << "Trap({0,1,0,2,1,0,1,3,2,1,2,1}): ";
  {
    auto result = Trap({0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1});
    std::cout << result << "\n";
  }

  return 0;
}

File: Java Playground/05_TwoPointers.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Two-Pointer Template & Techniques.
 *
 * <p>Two pointers is a technique where two pointers traverse an array/string from
 * different positions based on a condition. Key patterns:
 *
 * <p><b>OPPOSITE DIRECTION:</b> Initialize pointers at both ends, converge toward
 * center. Best for sorted arrays, palindromes, container problems. Condition typically
 * checks sum or comparison at current positions.
 *
 * <p><b>SAME DIRECTION:</b> Fast pointer explores, slow pointer builds or maintains
 * result. Common for removals (remove duplicates, remove element), moves (move zeroes).
 * Slower pointer always ≤ faster pointer.
 *
 * <p><b>PARTITION:</b> Rearrange in-place using pointers to mark regions (Dutch Flag).
 *
 * <p><b>VS SLIDING WINDOW:</b> Two pointers ideal for sorted input requiring fixed
 * operations. Sliding window for contiguous subarrays with dynamic window size.
 *
 * <p>Time: O(n) for single pass, O(n log n) with pre-sort. Space: O(1) in-place or
 * O(n) for output.
 */
public class TwoPointers {

  /**
   * Two Sum - Sorted Array.
   *
   * @param numbers sorted array (1-indexed in LeetCode, 0-indexed here)
   * @param target sum to find
   * @return int[] with two indices that sum to target [idx1, idx2] or [-1, -1]
   *     if not found
   * @throws NullPointerException if numbers is null
   *
   *         <p>LeetCode 167 (Medium). Time: O(n). Space: O(1).
   *
   *         <p>Opposite direction: left at start, right at end. If sum too small,
   *         move left right. If sum too large, move right left.
   */
  public static int[] twoSumSorted(int[] numbers, int target) {
    if (numbers == null || numbers.length < 2) {
      return new int[]{-1, -1};
    }

    int left = 0;
    int right = numbers.length - 1;

    while (left < right) {
      int sum = numbers[left] + numbers[right];
      if (sum == target) {
        return new int[]{left + 1, right + 1}; // 1-indexed for LeetCode
      } else if (sum < target) {
        left++;
      } else {
        right--;
      }
    }

    return new int[]{-1, -1};
  }

  /**
   * Three Sum - Find all unique triplets summing to zero.
   *
   * @param nums input array (may contain duplicates)
   * @return List of lists, each containing 3 integers that sum to 0, no duplicates
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 15 (Medium). Time: O(n²). Space: O(1) excluding output.
   *
   *         <p>Sort first O(n log n). For each element, use two-pointer on
   *         remaining to find pairs. Skip duplicates carefully.
   */
  public static List<List<Integer>> threeSum(int[] nums) {
    List<List<Integer>> result = new ArrayList<>();

    if (nums == null || nums.length < 3) {
      return result;
    }

    Arrays.sort(nums);

    for (int i = 0; i < nums.length - 2; i++) {
      // Optimization: if nums[i] > 0, no triplet sums to 0
      if (nums[i] > 0) {
        break;
      }

      // Skip duplicate first element
      if (i > 0 && nums[i] == nums[i - 1]) {
        continue;
      }

      int left = i + 1;
      int right = nums.length - 1;
      int target = -nums[i];

      while (left < right) {
        int sum = nums[left] + nums[right];
        if (sum == target) {
          result.add(Arrays.asList(nums[i], nums[left], nums[right]));

          // Skip duplicates for left
          while (left < right && nums[left] == nums[left + 1]) {
            left++;
          }
          // Skip duplicates for right
          while (left < right && nums[right] == nums[right - 1]) {
            right--;
          }

          left++;
          right--;
        } else if (sum < target) {
          left++;
        } else {
          right--;
        }
      }
    }

    return result;
  }

  /**
   * Four Sum - Find all unique quadruplets summing to target.
   *
   * @param nums input array
   * @param target sum to find
   * @return List of lists, each containing 4 integers that sum to target
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 18 (Medium). Time: O(n³). Space: O(1) excluding output.
   *
   *         <p>Sort, then nested loops. For each pair (i, j), use two-pointer on
   *         remaining. Watch for integer overflow with target.
   */
  public static List<List<Integer>> fourSum(int[] nums, int target) {
    List<List<Integer>> result = new ArrayList<>();

    if (nums == null || nums.length < 4) {
      return result;
    }

    Arrays.sort(nums);

    for (int i = 0; i < nums.length - 3; i++) {
      // Skip duplicate outer element
      if (i > 0 && nums[i] == nums[i - 1]) {
        continue;
      }

      // Optimization: if nums[i] * 4 > target, no quadruplet possible
      if ((long) nums[i] * 4 > target) {
        break;
      }

      for (int j = i + 1; j < nums.length - 2; j++) {
        // Skip duplicate inner element
        if (j > i + 1 && nums[j] == nums[j - 1]) {
          continue;
        }

        // Optimization: if nums[i] + nums[j] * 3 > target, no quadruplet
        if ((long) nums[i] + (long) nums[j] * 3 > target) {
          break;
        }

        int left = j + 1;
        int right = nums.length - 1;
        long twoSum = (long) target - nums[i] - nums[j];

        while (left < right) {
          long sum = (long) nums[left] + nums[right];
          if (sum == twoSum) {
            result.add(
                Arrays.asList(nums[i], nums[j], nums[left], nums[right]));

            // Skip duplicates
            while (left < right && nums[left] == nums[left + 1]) {
              left++;
            }
            while (left < right && nums[right] == nums[right - 1]) {
              right--;
            }

            left++;
            right--;
          } else if (sum < twoSum) {
            left++;
          } else {
            right--;
          }
        }
      }
    }

    return result;
  }

  /**
   * Container With Most Water - Find max area between two vertical lines.
   *
   * @param height array where height[i] is the vertical line at position i
   * @return maximum area of water that can be contained
   * @throws NullPointerException if height is null
   *
   *         <p>LeetCode 11 (Medium). Time: O(n). Space: O(1).
   *
   *         <p>Opposite direction: start with widest container, greedily shrink
   *         from the side with smaller height (it's the limiting factor).
   */
  public static int containerWithMostWater(int[] height) {
    if (height == null || height.length < 2) {
      return 0;
    }

    int maxArea = 0;
    int left = 0;
    int right = height.length - 1;

    while (left < right) {
      int width = right - left;
      int currentHeight = Math.min(height[left], height[right]);
      int area = width * currentHeight;
      maxArea = Math.max(maxArea, area);

      // Move the pointer pointing to smaller height (to find potentially taller
      // line)
      if (height[left] < height[right]) {
        left++;
      } else {
        right--;
      }
    }

    return maxArea;
  }

  /**
   * Remove Duplicates from Sorted Array (in-place).
   *
   * @param nums sorted array with possible duplicates
   * @return count of unique elements; first k elements contain unique values
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 26 (Easy). Time: O(n). Space: O(1).
   *
   *         <p>Same direction: slow pointer (write position), fast pointer
   *         (exploration). When nums[fast] != nums[slow], increment slow and
   *         copy nums[fast] to nums[slow].
   */
  public static int removeDuplicates(int[] nums) {
    if (nums == null || nums.length == 0) {
      return 0;
    }

    int slow = 0;

    for (int fast = 1; fast < nums.length; fast++) {
      if (nums[fast] != nums[slow]) {
        slow++;
        nums[slow] = nums[fast];
      }
    }

    return slow + 1;
  }

  /**
   * Remove Element (in-place).
   *
   * @param nums input array
   * @param val value to remove
   * @return count of elements not equal to val; first k elements don't contain
   *     val
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 27 (Easy). Time: O(n). Space: O(1).
   *
   *         <p>Same direction: slow tracks position for non-val elements. When
   *         nums[fast] != val, copy to nums[slow] and increment slow.
   */
  public static int removeElement(int[] nums, int val) {
    if (nums == null) {
      return 0;
    }

    int slow = 0;

    for (int fast = 0; fast < nums.length; fast++) {
      if (nums[fast] != val) {
        nums[slow] = nums[fast];
        slow++;
      }
    }

    return slow;
  }

  /**
   * Move Zeroes (in-place, maintain relative order).
   *
   * @param nums input array
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 283 (Easy). Time: O(n). Space: O(1).
   *
   *         <p>Same direction: slow tracks non-zero position. Move all non-zeros
   *         to front, then fill rest with zeros.
   */
  public static void moveZeroes(int[] nums) {
    if (nums == null) {
      return;
    }

    int slow = 0;

    // Move all non-zero elements to the front
    for (int fast = 0; fast < nums.length; fast++) {
      if (nums[fast] != 0) {
        nums[slow] = nums[fast];
        slow++;
      }
    }

    // Fill remaining with zeros
    while (slow < nums.length) {
      nums[slow] = 0;
      slow++;
    }
  }

  /**
   * Sort Colors (Dutch Flag problem, in-place).
   *
   * @param nums array containing only 0, 1, 2 (representing red, white, blue)
   * @throws NullPointerException if nums is null
   *
   *         <p>LeetCode 75 (Medium). Time: O(n). Space: O(1).
   *
   *         <p>Partition using three pointers: left (0s boundary), right (2s
   *         boundary), mid (current). Swap to maintain invariant:
   *         [0..left-1] = 0, [left..mid-1] = 1, [right+1..end] = 2.
   */
  public static void sortColors(int[] nums) {
    if (nums == null) {
      return;
    }

    int left = 0;   // boundary for 0s
    int mid = 0;    // current position
    int right = nums.length - 1; // boundary for 2s

    while (mid <= right) {
      if (nums[mid] == 0) {
        swap(nums, left, mid);
        left++;
        mid++;
      } else if (nums[mid] == 1) {
        mid++;
      } else { // nums[mid] == 2
        swap(nums, mid, right);
        right--;
      }
    }
  }

  /**
   * Valid Palindrome II (allow one deletion).
   *
   * @param s input string
   * @return true if string is palindrome or can be made palindrome by deleting
   *     at most one character
   * @throws NullPointerException if s is null
   *
   *         <p>LeetCode 680 (Medium). Time: O(n). Space: O(1).
   *
   *         <p>Opposite direction: converge from both ends. On mismatch, try
   *         deleting either left or right character and check if remaining is
   *         valid palindrome.
   */
  public static boolean validPalindrome(String s) {
    if (s == null) {
      return false;
    }

    int left = 0;
    int right = s.length() - 1;

    while (left < right) {
      if (s.charAt(left) != s.charAt(right)) {
        // Try deleting left or right character
        return isPalindrome(s, left + 1, right) ||
               isPalindrome(s, left, right - 1);
      }
      left++;
      right--;
    }

    return true;
  }

  /**
   * Helper: check if s[left..right] is valid palindrome.
   *
   * @param s input string
   * @param left left boundary
   * @param right right boundary
   * @return true if s[left..right] is palindrome
   */
  private static boolean isPalindrome(String s, int left, int right) {
    while (left < right) {
      if (s.charAt(left) != s.charAt(right)) {
        return false;
      }
      left++;
      right--;
    }
    return true;
  }

  /**
   * Trap Rainwater (two-pointer approach).
   *
   * @param height array of bar heights
   * @return total water trapped between bars
   * @throws NullPointerException if height is null
   *
   *         <p>LeetCode 42 (Hard). Time: O(n). Space: O(1).
   *
   *         <p>Key insight: water at position i depends on max height to left
   *         and right. Use two pointers from both ends, maintain max seen so far
   *         on each side. Move pointer pointing to smaller height (limits water
   *         at that position).
   */
  public static int trap(int[] height) {
    if (height == null || height.length == 0) {
      return 0;
    }

    int left = 0;
    int right = height.length - 1;
    int leftMax = 0;
    int rightMax = 0;
    int water = 0;

    while (left < right) {
      if (height[left] < height[right]) {
        if (height[left] >= leftMax) {
          leftMax = height[left];
        } else {
          water += leftMax - height[left];
        }
        left++;
      } else {
        if (height[right] >= rightMax) {
          rightMax = height[right];
        } else {
          water += rightMax - height[right];
        }
        right--;
      }
    }

    return water;
  }

  /**
   * Helper: swap two elements in array.
   *
   * @param arr input array
   * @param i first index
   * @param j second index
   */
  private static void swap(int[] arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  /**
   * Main method demonstrating all two-pointer techniques.
   */
  public static void main(String[] args) {
    System.out.println("=== TWO POINTERS TECHNIQUES ===\n");

    // Two Sum Sorted
    int[] numbers = {2, 7, 11, 15};
    int[] result = twoSumSorted(numbers, 9);
    System.out.println("Two Sum (Sorted): " + Arrays.toString(result));

    // Three Sum
    int[] nums = {-1, 0, 1, 2, -1, -4};
    List<List<Integer>> triplets = threeSum(nums);
    System.out.println("Three Sum: " + triplets);

    // Four Sum
    int[] nums2 = {1000000000, 1000000000, 1000000000, 1000000000};
    List<List<Integer>> quadruplets = fourSum(nums2, -294967296);
    System.out.println("Four Sum: " + quadruplets);

    // Container With Most Water
    int[] height1 = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    System.out.println("Container Max Area: " + containerWithMostWater(height1));

    // Remove Duplicates
    int[] arr1 = {1, 1, 2};
    int len = removeDuplicates(arr1);
    System.out.println("Remove Duplicates: length=" + len + ", arr=" +
        Arrays.toString(Arrays.copyOf(arr1, len)));

    // Remove Element
    int[] arr2 = {3, 2, 2, 3};
    int len2 = removeElement(arr2, 3);
    System.out.println("Remove Element (val=3): length=" + len2);

    // Move Zeroes
    int[] arr3 = {0, 1, 0, 3, 12};
    moveZeroes(arr3);
    System.out.println("Move Zeroes: " + Arrays.toString(arr3));

    // Sort Colors
    int[] arr4 = {2, 0, 2, 1, 1, 0};
    sortColors(arr4);
    System.out.println("Sort Colors: " + Arrays.toString(arr4));

    // Valid Palindrome II
    String s1 = "abca";
    System.out.println("Valid Palindrome II (abca): " + validPalindrome(s1));

    // Trap Rainwater
    int[] height2 = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    System.out.println("Trap Rainwater: " + trap(height2));
  }
}
